### ISSUE TITLE
DBusException on import (AccessDenied and NoReply) on certain environment

### ISSUE THREAD

**Issue by JonnyJD** [jaraco/keyring#183]
The issue was originally reported for packaging of isrcsubmit in JonnyJD/musicbrainz-isrcsubmit#108.

The problem is that `import keyring` fails when run by a user with an environment "inherited" by another user. The user does not have access to the dbus session of the original user and that breaks keyring already at the import.

The "inherited" environment is wrong, but keyring should not care about that at `import`.

The version of keyring I am running is 6.1.1 and python-secretstorage 2.1.3.

An example failing environmet (when running `import keyring` with an id of 1007):

```
DISPLAY=:0.1
DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
```

The error messages were different for 2 different users. For one user (@Freso) only `DISPLAY` was relevant and the error message on import was:
`dbus.exceptions.DBusException: org.freedesktop.DBus.Error.NoReply: Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.`

For my the dbus session variable was relevant and I got this error:
`dbus.exceptions.DBusException: org.freedesktop.DBus.Error.AccessDenied: Failed to connect to socket /run/user/1000/bus: Permission denied`. I do have to unset `DISPLAY`, too. Otherwise the import takes several seconds.

**Comment by JonnyJD** [comment]
The full trace is in my case:

```
>>> import keyring
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.5/site-packages/keyring/__init__.py", line 12, in <module>
    from .core import (set_keyring, get_keyring, set_password, get_password,
  File "/usr/lib/python3.5/site-packages/keyring/core.py", line 158, in <module>
    init_backend()
  File "/usr/lib/python3.5/site-packages/keyring/core.py", line 58, in init_backend
    set_keyring(load_config() or _get_best_keyring())
  File "/usr/lib/python3.5/site-packages/keyring/core.py", line 66, in _get_best_keyring
    keyrings = backend.get_all_keyring()
  File "/usr/lib/python3.5/site-packages/keyring/util/__init__.py", line 20, in wrapper
    func.always_returns = func(*args, **kwargs)
  File "/usr/lib/python3.5/site-packages/keyring/backend.py", line 180, in get_all_keyring
    exceptions=TypeError))
  File "/usr/lib/python3.5/site-packages/keyring/util/__init__.py", line 29, in suppress_exceptions
    for callable in callables:
  File "/usr/lib/python3.5/site-packages/keyring/backend.py", line 172, in is_class_viable
    keyring_cls.priority
  File "/usr/lib/python3.5/site-packages/keyring/util/properties.py", line 22, in __get__
    return self.fget.__get__(None, owner)()
  File "/usr/lib/python3.5/site-packages/keyring/util/XDG.py", line 18, in wrapper
    return func(*args, **kwargs) * self.multiplier
  File "/usr/lib/python3.5/site-packages/keyring/backends/kwallet.py", line 136, in priority
    bus = dbus.SessionBus()
  File "/usr/lib/python3.5/site-packages/dbus/_dbus.py", line 211, in __new__
    mainloop=mainloop)
  File "/usr/lib/python3.5/site-packages/dbus/_dbus.py", line 100, in __new__
    bus = BusConnection.__new__(subclass, bus_type, mainloop=mainloop)
  File "/usr/lib/python3.5/site-packages/dbus/bus.py", line 122, in __new__
    bus = cls._new_for_bus(address_or_type, mainloop=mainloop)
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.AccessDenied: Failed to connect to socket /run/user/1000/bus: Permission denied
```

I have kwallet 5.17 installed and I am using (KDE) Plasma 5, if that helps or makes a difference. I am not actively using kwallet, but I think I didn't have a real choice other than creating a wallet when starting the first KDE app in Plasma 5 (or maybe when I tried to fiddle with akonadi).
The point is: keyring shouldn't need to care about any of this on the `import keyring` step.

EDIT: I might note I am not "really" using Plasma 5 as a desktop environment. I am using fluxbox. I do hower use applications that are part of KDE.

**Comment by jaraco** [comment]
The reason DBus is invoked and access attempted during import is because each of the backends are expected to probe the environment to determine which backends are viable. I agree there's a bug here in that it should never fail or block for long periods during import.

**Comment by mitya57** [comment]
> `dbus.exceptions.DBusException: org.freedesktop.DBus.Error.AccessDenied: Failed to connect to socket /run/user/1000/bus: Permission denied`

As I understand, this should be fixed with 6dba134488ba4c1d056d979e7bf058fe6f659213 (i.e. in 7.0), as Keyring will now intercept this `DBusException` and turn it into a `RuntimeError` (which will make the KWallet backend not viable).

Also this bug illustrates one more reason for why presence of env variables is a bad indicator for D-Bus availability.

**Comment by jaraco** [comment]
Oh! Maybe. I thought the reported issue was related to Gnome keyring which is still using the old env var check and not trapping all errors. If the OP can test with 7.0, that'd be great.

**Comment by Freso** [comment]
FWIW, my testing environment is using GNOME's keyring. I'll try to see if I can get around to testing the latest -git keyring module.

**Comment by mitya57** [comment]
GNOME Keyring works without `$DISPLAY` as well (just checked that), so I've just submitted #185 to update it.

**Comment by Freso** [comment]
My PKGBUILD builds fine with v7.0. Thanks! @JonnyJD, does it work for you now?

**Comment by jaraco** [comment]
Between 7.0 and 7.0.1, I believe the issue should be fixed. Please re-open (or request such) if not.

**Comment by JonnyJD** [comment]
The problem with $DISPLAY is resolved, but the problem with DBUS_SESSION_BUS_ADDRESS still exists with python-keyring 7.0.1
Meaning I still get `dbus.exceptions.DBusException: org.freedesktop.DBus.Error.AccessDenied: Failed to connect to socket /run/user/1000/bus: Permission denied`

However, the stacktrace changed and now it looks like it fails in the SecretService/secretstorage backend (with the same error message as before in kwallet):

```
$ python -c "import keyring"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.5/site-packages/keyring/__init__.py", line 12, in <module>
    from .core import (set_keyring, get_keyring, set_password, get_password,
  File "/usr/lib/python3.5/site-packages/keyring/core.py", line 173, in <module>
    init_backend()
  File "/usr/lib/python3.5/site-packages/keyring/core.py", line 59, in init_backend
    set_keyring(load_config() or _get_best_keyring())
  File "/usr/lib/python3.5/site-packages/keyring/core.py", line 83, in _get_best_keyring
    keyrings = backend.get_all_keyring()
  File "/usr/lib/python3.5/site-packages/keyring/util/__init__.py", line 20, in wrapper
    func.always_returns = func(*args, **kwargs)
  File "/usr/lib/python3.5/site-packages/keyring/backend.py", line 180, in get_all_keyring
    exceptions=TypeError))
  File "/usr/lib/python3.5/site-packages/keyring/util/__init__.py", line 29, in suppress_exceptions
    for callable in callables:
  File "/usr/lib/python3.5/site-packages/keyring/backend.py", line 172, in is_class_viable
    keyring_cls.priority
  File "/usr/lib/python3.5/site-packages/keyring/util/properties.py", line 22, in __get__
    return self.fget.__get__(None, owner)()
  File "/usr/lib/python3.5/site-packages/keyring/util/XDG.py", line 18, in wrapper
    return func(*args, **kwargs) * self.multiplier
  File "/usr/lib/python3.5/site-packages/keyring/backends/SecretService.py", line 31, in priority
    bus = secretstorage.dbus_init()
  File "/usr/lib/python3.5/site-packages/secretstorage/__init__.py", line 50, in dbus_init
    return dbus.SessionBus()
  File "/usr/lib/python3.5/site-packages/dbus/_dbus.py", line 211, in __new__
    mainloop=mainloop)
  File "/usr/lib/python3.5/site-packages/dbus/_dbus.py", line 100, in __new__
    bus = BusConnection.__new__(subclass, bus_type, mainloop=mainloop)
  File "/usr/lib/python3.5/site-packages/dbus/bus.py", line 122, in __new__
    bus = cls._new_for_bus(address_or_type, mainloop=mainloop)
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.AccessDenied: Failed to connect to socket /run/user/1000/bus: Permission denied
```

**Comment by mitya57** [comment]
I will fix that on SecretStorage side and roll a new release today.

**Comment by mitya57** [comment]
This is now fixed in SecretStorage 2.1.4.

**Comment by JonnyJD** [comment]
Thanks a lot. That also fixed my problems.
